### PR TITLE
init synchronous architecture

### PR DESCRIPTION
**Pull request by asamuj**

### CODE REVIEW

File: chain/node.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+)
+
+const (
+	chainStatusPath = "chain-status"
+	blockPath       = "block"
+)
+
+// Node gets the required data according to the HTTP request
+// from the photon node.
+type Node struct {
```
Comment (maxkaneco) [inline chain/node.go:21]: NodeClient

File: chain/node.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
+package chain
```
Comment (maxkaneco) [inline chain/node.go:1]: how about merging this into sync package to keep repo struct simple?

File: cmd/flags.go [hunk 3]
```diff
@@ -0,0 +1,26 @@
+package cmd
+
+import "github.com/urfave/cli/v2"
+
+var (
+	//ConfigPathFlag specifies the config file path.
+	ConfigPathFlag = &cli.StringFlag{
```
Comment (maxkaneco) [inline cmd/flags.go:7]: move into cmd/sync and make it unexported

File: cmd/runtime/version/version.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
+// Package version executes and returns the version string
```
Comment (maxkaneco) [inline cmd/runtime/version/version.go:1]: move into to cmd/sync pkg

File: cmd/sync/main.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
+package main
```
Comment (maxkaneco) [inline cmd/sync/main.go:1]: Let's call it indexer?

File: config/sync_config.go [hunk 6]
```diff
@@ -0,0 +1,10 @@
+package config
+
+import "github.com/photon-storage/photon-explorer/database/mysql"
+
+// SyncerConfig represent sync of config
+type SyncerConfig struct {
```
Comment (maxkaneco) [inline config/sync_config.go:6]: IndexerConfig defines the config for indexer service.
Move to cmd/sync/main.go

File: config/config.go [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+import (
+	"encoding/json"
+	"os"
+
+	"github.com/pkg/errors"
+)
+
+// LoadConfig reads path file to the config object.
+func LoadConfig(filePath string, config interface{}) error {
```
Comment (maxkaneco) [inline config/config.go:11]: Load()

File: config/sync_config.go [hunk 8]
```diff
@@ -0,0 +1,10 @@
...
+	MySQL       mysql.Config `json:"mysql"`
+	SyncSeconds uint64       `json:"sync_seconds"`
+	NodeURL     string       `json:"node_url"`
```
Comment (maxkaneco) [inline config/sync_config.go:9]: NodeEndpoint

File: config/sync_config.go [hunk 9]
```diff
@@ -0,0 +1,10 @@
+package config
+
+import "github.com/photon-storage/photon-explorer/database/mysql"
+
+// SyncerConfig represent sync of config
+type SyncerConfig struct {
+	MySQL       mysql.Config `json:"mysql"`
+	SyncSeconds uint64       `json:"sync_seconds"`
```
Comment (maxkaneco) [inline config/sync_config.go:8]: RefreshInterval

File: config/sync_config_test.json [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+      "port": 3306,
+      "username": "root",
+      "password": "mysql123",
+      "db_name": "photon_explorer"
+    },
+    "conns": {
+      "max_open_conns": 40,
+      "max_idle_conns": 20
+    },
+    "log_level": 4
```
Comment (maxkaneco) [inline config/sync_config_test.json:14]: use string? who knows what 4 means

File: database/mysql/config.go [hunk 11]
```diff
@@ -0,0 +1,22 @@
+package mysql
+
+// Config represent root of mysql config
+type Config struct {
```
Comment (maxkaneco) [inline database/mysql/config.go:4]: Config defines mysql configuration.

File: database/mysql/config.go [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+type connection struct {
+	Host     string `json:"host"`
+	Port     uint   `json:"port"`
+	Username string `json:"username"`
+	Password string `json:"password"`
+	DBName   string `json:"db_name"`
+}
+
+type connCfg struct {
+	MaxOpenConns int `json:"max_open_conns"`
```
Comment (maxkaneco) [inline database/mysql/config.go:20]: define in Config struct directly?

File: database/mysql/mysql.go [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+			slave.Password,
+			slave.Host,
+			slave.Port,
+			slave.DBName,
+		)
+		slaveDSNs = append(slaveDSNs, mysql.Open(slaveDSN))
+	}
+
+	dbResolverCfg := dbresolver.Config{
+		Sources:  []gorm.Dialector{mysql.Open(masterDSN)},
```
Comment (maxkaneco) [inline database/mysql/mysql.go:44]: do we need to open master again here? Use db?

File: sync/block.go [hunk 14]
```diff
@@ -0,0 +1,11 @@
+package sync
```
Comment (maxkaneco) [inline sync/block.go]: lets call it indexer

File: sync/sync.go [hunk 15]
```diff
@@ -0,0 +1,97 @@
+package sync
```
Comment (maxkaneco) [inline sync/sync.go]: eventprocessor.go

File: sync/sync.go [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+
+// Run executing the timing task of processing chain data.
+func (e *EventProcessor) Run(ctx context.Context) {
+	ticker := time.NewTicker(time.Duration(e.syncSeconds) * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			localSlot, localHash, err := e.localChainStatus()
```
Comment (maxkaneco) [inline sync/sync.go]: I dont think we need to query DB again and again. We know the chain status from the last update.

File: chain/node.go (hunk 2 above)
Comment (dorisRiver) [reply chain/node.go:1]: api and indexer are two independent services, and api also needs the node service.

File: cmd/flags.go (hunk 3 above)
Comment (dorisRiver) [reply cmd/flags.go:7]: api need this flag too.

File: cmd/runtime/version/version.go (hunk 4 above)
Comment (dorisRiver) [reply cmd/runtime/version/version.go:1]: api need version pkg too.

File: database/mysql/mysql.go (hunk 13 above)
Comment (dorisRiver) [reply database/mysql/mysql.go:44]: db doesn't implement GORM database dialector interface.

**Review by maxkaneco — APPROVED** [review]

**Comment by maxkaneco** [comment]
PPH=15hrs
including design work

**Pull request merged**